assets/mdk/brittleness.mdk

```diff
@@ -0,0 +1,179 @@
+
+A nice coincidence is that programs written according to these principles tend to be more understandable to humans, as well. Reducing the work a prover needs to do when reasoning about a program has the effect of reducing the amount of thinking a human has to do to understand what a program does, and why it’s correct. Constructing small, relatively isolated components relates strongly to the concepts of [cohesion](https://en.wikipedia.org/wiki/Cohesion_(computer_science)) and [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) used in software engineering.
+
+Those principles are rather general, however, and there are several more specific guidelines for how to apply them in practice in Dafny.
+
+* Prefer `opaque` functions, with explicit `reveal` statements where necessary.
+* Avoid `requires` and `ensures` clauses on functions, except when necessary to make them total, and prefer a collection of `lemma`s, each of which establishes a single property of the function.
+* Avoid lemmas that establish universally-qualified facts, preferring an additional parameter for each variable that would otherwise be quantified. It’s possible to wrap such a lemma to use universal quantification if it’s necessary, but many uses can pass in specific instantiations.
+* Avoid subset types except perhaps for simple integer ranges.
+* Prove the correctness of each method by including a single postcondition that establishes equivalence with a function or agreement with a relation. Other properties can be proved as independent lemmas about that function or relation.
+
+The Dafny-VMC project includes some [more detailed guidelines](https://github.com/dafny-lang/Dafny-VMC/blob/main/Guidelines.md). Sometimes it can be difficult to re-architect large, existing systems to follow those guidelines strictly, and we have some additional documentation on [optimizing verification](https://dafny.org/latest/VerificationOptimization/VerificationOptimization) that includes techniques that can help in this case.
+
+# Conclusion
+
+Brittleness is an inevitable consequence of the combination of expressiveness and automation available in Dafny. However, although it is unavoidable in theory, it can be dramatically reduced in practice. Aiming to reduce resource use, and setting limits on it as part of your build process, is likely the single most effective way to do so. This can be achieved by writing code in small, self-contained units that each have a clear, single purpose and don’t leak implementation details except when necessary. Within these units, additional details about intermediate proof steps can help. Although `assert` and `calc` statements are the most straightforward way to do this, using the structure of [natural deduction](https://en.wikipedia.org/wiki/Natural_deduction) can be even more effective.
```

Comment:
Should we add after the first sentence that an important part of good verified software engineering is to work on improving the proofs? I think we should.

Reply:
Ah, here comes natural deduction :)